hasktorch/src/Torch/Traversable.hs

```diff
@@ -0,0 +1,117 @@
+{-# LANGUAGE DefaultSignatures #-}
+{-# LANGUAGE RankNTypes #-}
+{-# LANGUAGE TypeOperators #-}
+{-# LANGUAGE MultiParamTypeClasses #-}
+{-# LANGUAGE FlexibleInstances #-}
+{-# LANGUAGE FlexibleContexts #-}
+{-# LANGUAGE UndecidableInstances #-}
+{-# LANGUAGE TypeFamilies #-}
+{-# LANGUAGE DataKinds #-}
+
+module Torch.Traversable where
+
+import Control.Monad.State.Strict
+import GHC.Generics
+import Data.Foldable (toList)
+import Torch.Scalar
+
+class GTraversable a f where
+  gflatten :: f -> [a]
+  default gflatten :: (Generic f, GGTraversable a (Rep f)) => f -> [a]
+  gflatten f = gflatten' (from f)
+
+  gupdate :: f -> State [a] f
+  default gupdate :: (Generic f, GGTraversable a (Rep f)) => f -> State [a] f
+  gupdate f = to <$> gupdate' (from f)
+
+class GGTraversable a f where
+  gflatten' :: forall p. f p -> [a]
+  gupdate' :: forall p. f p -> State [a] (f p)
+
+instance GGTraversable a U1 where
+  gflatten' U1 = []
+  gupdate' U1 = return U1
+
+instance (GGTraversable a f, GGTraversable a g) => GGTraversable a (f :+: g) where
+  gflatten' (L1 x) = gflatten' x
+  gflatten' (R1 x) = gflatten' x
+  gupdate' (L1 x) = do
+    x' <- gupdate' x
+    return $ L1 x'
+  gupdate' (R1 x) = do
+    x' <- gupdate' x
+    return $ R1 x'
+
+instance (GGTraversable a f, GGTraversable a g) => GGTraversable a (f :*: g) where
+  gflatten' (x :*: y) = gflatten' x ++ gflatten' y
+  gupdate' (x :*: y) = do
+    x' <- gupdate' x
+    y' <- gupdate' y
+    return $ x' :*: y'
+
+instance (GTraversable a c) => GGTraversable a (K1 i c) where
+  gflatten' (K1 x) = gflatten x
+  gupdate' (K1 x) = do
+    x' <- gupdate x
+    return $ K1 x'
+
+instance (GGTraversable a f) => GGTraversable a (M1 i t f) where
+  gflatten' (M1 x) = gflatten' x
+  gupdate' (M1 x) = do
+    x' <- gupdate' x
+    return $ M1 x'
+
+instance GTraversable a a where
+  gflatten = pure
+  gupdate _ = gpop
+
+instance {-# OVERLAPS #-} (Generic f, GGTraversable a (Rep f)) => GTraversable a f
```

Comment:
I'm not sure about this. How is the user experience if
1. No Generic instance exists  for a given f?
2. No Generic instance can be automatically derived for a given f?
3. A user wishes to implement a custom GTraversable instance for a given f?

Reply:
> 1. No Generic instance exists for a given f?

f can become a GTraversable instance without a Generic instance.

> 2. No Generic instance can be automatically derived for a given f?

We can use a custom Generic instance for a given f.

> 3. A user wishes to implement a custom GTraversable instance for a given f?

A user can implement a custom GTraversable instance of f without a Generic instance.
A user can not implement a custom GTraversable instance of f with a Generic instance.

Oh, we can not use both a custom Generic instance and a custom GTraversable instance for a given f.

Reply:
I will make some examples for the instances.